### PR TITLE
Enforce git secrets are configured before allowing development [BA-5810]

### PR DESCRIPTION
**Pull request by cjllanwarne**
Where "enforce" means "don't allow Cromwell to compile unless you have the hooks set up."

Workaround is running `sbt -Dignore-hooks-check=true`

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -1,5 +1,28 @@
...
+  val checkHooksDir: Seq[String] = Seq("bash", "-c", "git config core.hooksPath | grep -q 'hooks/'")
+  s.log.info("Executing pre-compile script...")
+  if((checkHooksDir !) == 0) {
+    s.log.success("Pre compile script ran successfully!")
+  } else {
+    s.log.error("Pre compile script failed.")
+    s.log.error("Git hooks directory is not configured.")
+    s.log.error("Run before continuing: git config --add core.hooksPath hooks/")
+    System.exit(1)
+//    throw new sbt.MessageOnlyException("Git hooks directory is not configured. Run 'git config --add core.hooksPath hooks/' before continuing.")
```
Comment (cjllanwarne) [inline build.sbt]: For some reason `throw` did not halt the process, which is why I had to resort to `s.log.error` and `System.exit(1)` to actually make sbt quit

File: build.sbt [hunk 2]
```diff
@@ -1,5 +1,49 @@
...
+import com.typesafe.config.ConfigFactory
+
+import scala.sys.process._
+
+// Pre-sbt checks:
+
+lazy val preCompile = taskKey[Unit]("Execute pre-compile scripts")
+
+preCompile := {
+  val s: TaskStreams = streams.value
```
Comment (kshakir) [inline build.sbt]:
```
  val l = streams.value.log
```

Then use the logger everywhere instead of `s.log`.

File: build.sbt [hunk 3]
```diff
@@ -1,5 +1,49 @@
...
+  s.log.info("Executing pre-compile script...")
+
+  def checkHooksDir() = Seq("bash", "-c", "git config core.hooksPath | grep -q 'hooks/'").!
+  def configureHooks() = configureHooksCommand.!
+
+  def ignoreHookCheck: Boolean = {
+    val config = ConfigFactory.load()
+    if (config.hasPath("ignore-hooks-check")) {
+      config.getBoolean("ignore-hooks-check")
+    } else false
```
Comment (kshakir) [inline build.sbt]:
```scala
sys.props.get("ignore-hooks-check").forall(_.toBoolean)
```

Lighter weight supports the same documented `-D` syntax.

File: build.sbt [hunk 4]
```diff
@@ -1,5 +1,49 @@
...
+    s.log.success("Pre compile script ran successfully")
+  } else {
+    val message = s"You are not running our custom git commit hooks. If you are making changes to the codebase, we recommend doing this (by running '$configureHooksCommandString') to ensure that your cryptographic secrets are not committed to our repository by accident."
+    
+    if (ignoreHookCheck) {
+      s.log.warn(message)
+    } else {
+      s.log.error(message)
+      s.log.error("If you don't want to set up hooks (if you never intend to commit to the cromwell repo, can be sure that you won't commit secrets by accident, or have already installed git-secrets in this repo separately), you can suppress this error by running with: 'sbt -Dignore-hooks-check=true [...]'")
+      System.exit(1)
```
Comment (kshakir) [inline build.sbt]: `sys.exit(1)` is slightly more scala-esque

**Comment by danbills** [comment]
I'm a :-1: on this idea.  Seems like it is overly demanding as compilation is less of an ask than committing code.   For instance, now our build servers must have git secrets installed where it is irrelevant.

**Comment by cjllanwarne** [comment]
@aednichols tested that just now. The experience is similar to the "I don't have git hooks installed" case (ie see the two final `[error]` messages):
```
$ sbt compile
[info] Loading settings for project global-plugins from idea.sbt ...
[info] Loading global plugins from /Users/chrisl/.sbt/1.0/plugins
[info] Loading project definition from /Users/chrisl/clonetest/cromwell-cjl_prerendered_secrets_hooks/project/project
[info] Updating ProjectRef(uri("file:/Users/chrisl/clonetest/cromwell-cjl_prerendered_secrets_hooks/project/project/"), "cromwell-cjl_prerendered_secrets_hooks-build-build")...
[info] Done updating.
[info] Loading settings for project cromwell-cjl_prerendered_secrets_hooks-build from plugins.sbt,swagger2markup.sbt ...
[info] Loading project definition from /Users/chrisl/clonetest/cromwell-cjl_prerendered_secrets_hooks/project
[info] Updating ProjectRef(uri("file:/Users/chrisl/clonetest/cromwell-cjl_prerendered_secrets_hooks/project/"), "cromwell-cjl_prerendered_secrets_hooks-build")...
[info] Done updating.
[warn] There may be incompatibilities among your library dependencies; run 'evicted' to see detailed eviction warnings.
[info] Compiling 8 Scala sources to /Users/chrisl/clonetest/cromwell-cjl_prerendered_secrets_hooks/project/target/scala-2.12/sbt-1.0/classes ...
[info] Done compiling.
[info] Loading settings for project root from build.sbt ...
[info] Resolving key references (36497 settings) ...
[info] Set current project to root (in build file:/Users/chrisl/clonetest/cromwell-cjl_prerendered_secrets_hooks/)
[info] Executing in batch mode. For better performance use sbt's shell
[info] Executing pre-compile script...
[error] You are not running our custom git commit hooks. If you are making changes to the codebase, we recommend doing this (by running 'git config --add core.hooksPath hooks/') to ensure that your cryptographic secrets are not committed to our repository by accident.
[error] If you don't want to set up hooks (if you never intend to commit to the cromwell repo, can be sure that you won't commit secrets by accident, or have already installed git-secrets in this repo separately), you can suppress this error by running with: 'sbt -Dignore-hooks-check=true [...]'
```

**Comment by cjllanwarne** [comment]
NB: the "workaround" of running `sbt -Dignore-hooks-check=true compile` still works in this case

**Comment by cjllanwarne** [comment]
> For instance, now our build servers must have git secrets installed where it is irrelevant.

@danbills That's not true - unless they want to commit code. This only asks them to configure a set of git hooks which they'll never end up using (or to add a compile time option which makes the compilation ignore the check).

**Comment by danbills** [comment]
@cjllanwarne I'm not going to hold up the PR.  I like that there is a flag but it still feels onerous.  I understand the argument that it's worth it, I just disagree with it.

**Comment by cjllanwarne** [comment]
Closing based on feedback.

**Pull request closed without merging**